<a href="https://colab.research.google.com/github/priyakumar88380/Healthcare-Patient-Billing-Analysis/blob/main/PRCL_0017_NoChurn_Telecom_Prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!pip install pymysql sqlalchemy -q

In [2]:
from google.colab import userdata
db_host = userdata.get('DB_HOST')
db_user = userdata.get('DB_USER')
db_pass = userdata.get('DB_PASS')
db_name = userdata.get('DB_NAME')

In [3]:
from sqlalchemy import create_engine
import pandas as pd
from urllib.parse import quote_plus

# quote_plus safely encodes special characters in the password (like $, !, @, &)
safe_pass = quote_plus(db_pass)

connection_string = f"mysql+pymysql://{db_user}:{safe_pass}@{db_host}/{db_name}"
engine = create_engine(connection_string)

query = "SELECT * FROM telecom_churn_data;"
df = pd.read_sql(query, engine)

print(df.shape)
df.head()

(243553, 14)


,customer_id,telecom_partner,gender,age,state,city,pincode,date_of_registration,num_dependents,estimated_salary,calls_made,sms_sent,data_used,churn
0,1,Reliance Jio,F,25,Karnataka,Kolkata,755597,2020-01-01,4,124962,44,45,-361,0
1,2,Reliance Jio,F,55,Mizoram,Mumbai,125926,2020-01-01,2,130556,62,39,5973,0
2,3,Vodafone,F,57,Arunachal Pradesh,Delhi,423976,2020-01-01,0,148828,49,24,193,1
3,4,BSNL,M,46,Tamil Nadu,Kolkata,522841,2020-01-01,1,38722,80,25,9377,1
4,5,BSNL,F,26,Tripura,Delhi,740247,2020-01-01,2,55098,78,15,1393,0


In [4]:
from sqlalchemy import inspect
inspector = inspect(engine)
print(inspector.get_table_names())

['telecom_churn_data']


In [5]:
!pip install pymysql --quiet

In [6]:
import pymysql
import pandas as pd

connection = pymysql.connect(
    host='18.136.157.135',
    user='dm_team3',
    password='DM!$!Team!27@9!20&',
    database='project_telecom'
)

query = "SELECT * FROM telecom_churn_data"
df = pd.read_sql(query, connection)

print("Shape:", df.shape)
df.head()

/tmp/ipykernel_5732/4182733064.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query, connection)


Shape: (243553, 14)


,customer_id,telecom_partner,gender,age,state,city,pincode,date_of_registration,num_dependents,estimated_salary,calls_made,sms_sent,data_used,churn
0,1,Reliance Jio,F,25,Karnataka,Kolkata,755597,2020-01-01,4,124962,44,45,-361,0
1,2,Reliance Jio,F,55,Mizoram,Mumbai,125926,2020-01-01,2,130556,62,39,5973,0
2,3,Vodafone,F,57,Arunachal Pradesh,Delhi,423976,2020-01-01,0,148828,49,24,193,1
3,4,BSNL,M,46,Tamil Nadu,Kolkata,522841,2020-01-01,1,38722,80,25,9377,1
4,5,BSNL,F,26,Tripura,Delhi,740247,2020-01-01,2,55098,78,15,1393,0


In [7]:
!pip install sqlalchemy pymysql --quiet

In [8]:
from sqlalchemy import create_engine
from urllib.parse import quote_plus
import pandas as pd

# quote_plus safely converts special characters into a URL-safe format
password = quote_plus("DM!$!Team!27@9!20&")
username = "dm_team3"
host = "18.136.157.135"
database = "project_telecom"

connection_string = f"mysql+pymysql://{username}:{password}@{host}/{database}"
engine = create_engine(connection_string)

query = "SELECT * FROM telecom_churn_data"
df = pd.read_sql(query, engine)

print("Shape:", df.shape)
df.head()

Shape: (243553, 14)


,customer_id,telecom_partner,gender,age,state,city,pincode,date_of_registration,num_dependents,estimated_salary,calls_made,sms_sent,data_used,churn
0,1,Reliance Jio,F,25,Karnataka,Kolkata,755597,2020-01-01,4,124962,44,45,-361,0
1,2,Reliance Jio,F,55,Mizoram,Mumbai,125926,2020-01-01,2,130556,62,39,5973,0
2,3,Vodafone,F,57,Arunachal Pradesh,Delhi,423976,2020-01-01,0,148828,49,24,193,1
3,4,BSNL,M,46,Tamil Nadu,Kolkata,522841,2020-01-01,1,38722,80,25,9377,1
4,5,BSNL,F,26,Tripura,Delhi,740247,2020-01-01,2,55098,78,15,1393,0


In [9]:
!pip install sqlalchemy pymysql --quiet

from sqlalchemy import create_engine
from urllib.parse import quote_plus
import pandas as pd

password = quote_plus("DM!$!Team!27@9!20&")
username = "dm_team3"
host = "18.136.157.135"
database = "project_telecom"

connection_string = f"mysql+pymysql://{username}:{password}@{host}/{database}"
engine = create_engine(connection_string)

query = "SELECT * FROM telecom_churn_data"
df = pd.read_sql(query, engine)

print("Shape:", df.shape)
df.head()

Shape: (243553, 14)


,customer_id,telecom_partner,gender,age,state,city,pincode,date_of_registration,num_dependents,estimated_salary,calls_made,sms_sent,data_used,churn
0,1,Reliance Jio,F,25,Karnataka,Kolkata,755597,2020-01-01,4,124962,44,45,-361,0
1,2,Reliance Jio,F,55,Mizoram,Mumbai,125926,2020-01-01,2,130556,62,39,5973,0
2,3,Vodafone,F,57,Arunachal Pradesh,Delhi,423976,2020-01-01,0,148828,49,24,193,1
3,4,BSNL,M,46,Tamil Nadu,Kolkata,522841,2020-01-01,1,38722,80,25,9377,1
4,5,BSNL,F,26,Tripura,Delhi,740247,2020-01-01,2,55098,78,15,1393,0


In [10]:
# list all tables in the project_telecom database
query = "SHOW TABLES;"
tables = pd.read_sql(query, engine)
print(tables)

  Tables_in_project_telecom
0        telecom_churn_data


In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 243553 entries, 0 to 243552
Data columns (total 14 columns):
 #   Column                Non-Null Count   Dtype 
---  ------                --------------   ----- 
 0   customer_id           243553 non-null  int64 
 1   telecom_partner       243553 non-null  object
 2   gender                243553 non-null  object
 3   age                   243553 non-null  int64 
 4   state                 243553 non-null  object
 5   city                  243553 non-null  object
 6   pincode               243553 non-null  int64 
 7   date_of_registration  243553 non-null  object
 8   num_dependents        243553 non-null  int64 
 9   estimated_salary      243553 non-null  int64 
 10  calls_made            243553 non-null  int64 
 11  sms_sent              243553 non-null  int64 
 12  data_used             243553 non-null  int64 
 13  churn                 243553 non-null  int64 
dtypes: int64(9), object(5)
memory usage: 26.0+ MB


In [12]:
print(df['churn'].value_counts())
print(df['churn'].value_counts(normalize=True) * 100)

churn
0    194726
1     48827
Name: count, dtype: int64
churn
0    79.952208
1    20.047792
Name: proportion, dtype: float64


In [13]:
print(df.isnull().sum())

customer_id             0
telecom_partner         0
gender                  0
age                     0
state                   0
city                    0
pincode                 0
date_of_registration    0
num_dependents          0
estimated_salary        0
calls_made              0
sms_sent                0
data_used               0
churn                   0
dtype: int64


The churn target showed moderate imbalance (80% retained, 20% churned) — less severe than a highly imbalanced problem, but still warranting evaluation beyond plain accuracy, using precision, recall, and AUC

In [14]:
print(df.groupby('telecom_partner')['churn'].mean() * 100)
print()
print(df.groupby('gender')['churn'].mean() * 100)

telecom_partner
Airtel          20.366144
BSNL            19.860679
Reliance Jio    20.015379
Vodafone        19.948357
Name: churn, dtype: float64

gender
F    20.296999
M    19.881214
Name: churn, dtype: float64


In [15]:
numeric_cols = ['age', 'estimated_salary', 'num_dependents', 'calls_made', 'sms_sent', 'data_used']

for col in numeric_cols:
    print(f"\n{col} — mean by churn status:")
    print(df.groupby('churn')[col].mean())


age — mean by churn status:
churn
0    46.070699
1    46.105167
Name: age, dtype: float64

estimated_salary — mean by churn status:
churn
0    85083.726934
1    84771.527495
Name: estimated_salary, dtype: float64

num_dependents — mean by churn status:
churn
0    1.999302
1    1.990313
Name: num_dependents, dtype: float64

calls_made — mean by churn status:
churn
0    48.985590
1    49.110083
Name: calls_made, dtype: float64

sms_sent — mean by churn status:
churn
0    23.968068
1    23.855019
Name: sms_sent, dtype: float64

data_used — mean by churn status:
churn
0    4992.107941
1    4997.485510
Name: data_used, dtype: float64


In [16]:
df['date_of_registration'] = pd.to_datetime(df['date_of_registration'])
df['tenure_days'] = (pd.Timestamp.today() - df['date_of_registration']).dt.days

print("tenure_days — mean by churn status:")
print(df.groupby('churn')['tenure_days'].mean())

tenure_days — mean by churn status:
churn
0    1783.290875
1    1781.839945
Name: tenure_days, dtype: float64


In [17]:
print("state — churn rate (%):")
print(df.groupby('state')['churn'].mean().sort_values(ascending=False) * 100)

state — churn rate (%):
state
Jharkhand            21.119360
Karnataka            20.712267
Mizoram              20.646795
Uttarakhand          20.415537
Himachal Pradesh     20.410044
Madhya Pradesh       20.381082
Assam                20.299520
Odisha               20.247934
Nagaland             20.228918
Rajasthan            20.217143
Manipur              20.113282
Telangana            20.105881
Uttar Pradesh        20.075498
Sikkim               20.071355
Haryana              20.034325
Gujarat              20.023229
Goa                  19.956292
Tamil Nadu           19.935765
Maharashtra          19.909297
Arunachal Pradesh    19.892657
Andhra Pradesh       19.852006
Meghalaya            19.748964
Bihar                19.633963
Kerala               19.530086
Tripura              19.480370
Chhattisgarh         19.430679
Punjab               19.425127
West Bengal          19.400373
Name: churn, dtype: float64


In [18]:
drop_cols = ['customer_id', 'date_of_registration', 'pincode']
X = df.drop(columns=drop_cols + ['churn'])
y = df['churn']

In [19]:
categorical_cols = ['telecom_partner', 'gender', 'state', 'city']

In [21]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
from sklearn.metrics import classification_report, roc_auc_score

In [22]:
drop_cols = ['customer_id', 'date_of_registration', 'pincode']
X = df.drop(columns=drop_cols + ['churn'])
y = df['churn']

categorical_cols = ['telecom_partner', 'gender', 'state', 'city']
numeric_cols = ['age', 'estimated_salary', 'num_dependents', 'calls_made', 'sms_sent', 'data_used', 'tenure_days']

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

preprocessor = ColumnTransformer(transformers=[
    ('num', StandardScaler(), numeric_cols),
    ('cat', OneHotEncoder(handle_unknown='ignore'), categorical_cols)
])

models = {
    'Logistic Regression': LogisticRegression(max_iter=1000, class_weight='balanced'),
    'Random Forest': RandomForestClassifier(n_estimators=100, random_state=42, class_weight='balanced'),
    'XGBoost': XGBClassifier(random_state=42, eval_metric='logloss', scale_pos_weight=4)
}

results = {}
for name, model in models.items():
    pipe = Pipeline([
        ('preprocessor', preprocessor),
        ('classifier', model)
    ])
    pipe.fit(X_train, y_train)
    y_pred = pipe.predict(X_test)
    y_proba = pipe.predict_proba(X_test)[:, 1]

    print(f"\n{'='*50}\n{name}\n{'='*50}")
    print(classification_report(y_test, y_pred, target_names=['No Churn', 'Churn']))
    auc = roc_auc_score(y_test, y_proba)
    print(f"AUC: {auc:.4f}")
    results[name] = auc

print("\n\nSummary — AUC by model:")
for name, auc in results.items():
    print(f"{name}: {auc:.4f}")


Logistic Regression
              precision    recall  f1-score   support

    No Churn       0.80      0.51      0.62     38946
       Churn       0.20      0.50      0.29      9765

    accuracy                           0.51     48711
   macro avg       0.50      0.50      0.45     48711
weighted avg       0.68      0.51      0.55     48711

AUC: 0.5060

Random Forest
              precision    recall  f1-score   support

    No Churn       0.80      1.00      0.89     38946
       Churn       0.00      0.00      0.00      9765

    accuracy                           0.80     48711
   macro avg       0.40      0.50      0.44     48711
weighted avg       0.64      0.80      0.71     48711

AUC: 0.5023


/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))



XGBoost
              precision    recall  f1-score   support

    No Churn       0.80      0.56      0.66     38946
       Churn       0.20      0.45      0.28      9765

    accuracy                           0.53     48711
   macro avg       0.50      0.50      0.47     48711
weighted avg       0.68      0.53      0.58     48711

AUC: 0.5046


Summary — AUC by model:
Logistic Regression: 0.5060
Random Forest: 0.5023
XGBoost: 0.5046


In [23]:
# Use the best-performing model (XGBoost) to generate churn probabilities for ALL customers
best_pipe = Pipeline([
    ('preprocessor', preprocessor),
    ('classifier', XGBClassifier(random_state=42, eval_metric='logloss', scale_pos_weight=4))
])
best_pipe.fit(X_train, y_train)

# Predict churn probability for the FULL dataset (not just test set) — this is what a retention team would use
df['churn_probability'] = best_pipe.predict_proba(X)[:, 1]

# Create risk buckets — simple, interpretable tiers for a business audience
def assign_risk(prob):
    if prob >= 0.6:
        return 'High'
    elif prob >= 0.3:
        return 'Medium'
    else:
        return 'Low'

df['CHURN-FLAG'] = df['churn_probability'].apply(assign_risk)

# Quick check of the distribution
print(df['CHURN-FLAG'].value_counts())
print("\nAverage actual churn rate within each risk tier:")
print(df.groupby('CHURN-FLAG')['churn'].mean())

CHURN-FLAG
Medium    225862
High        9828
Low         7863
Name: count, dtype: int64

Average actual churn rate within each risk tier:
CHURN-FLAG
High      0.545788
Low       0.038026
Medium    0.191108
Name: churn, dtype: float64


In [24]:
# CORRECT: only assign risk scores using out-of-sample predictions (test set)
X_test_copy = X_test.copy()
X_test_copy['churn_probability'] = best_pipe.predict_proba(X_test)[:, 1]
X_test_copy['CHURN-FLAG'] = X_test_copy['churn_probability'].apply(assign_risk)
X_test_copy['actual_churn'] = y_test.values

print(X_test_copy['CHURN-FLAG'].value_counts())
print("\nAverage actual churn rate within each risk tier (test set only):")
print(X_test_copy.groupby('CHURN-FLAG')['actual_churn'].mean())

CHURN-FLAG
Medium    45586
High       1703
Low        1422
Name: count, dtype: int64

Average actual churn rate within each risk tier (test set only):
CHURN-FLAG
High      0.199648
Low       0.205345
Medium    0.200347
Name: actual_churn, dtype: float64


# PRCL-0017: No-Churn Telecom — Project Report

**Team ID:** PTID-CDS-JUL-26-11119
**DataMites Certified Data Scientist (CDS) Program — Live/Client Project**

## 1. Introduction

This project is part of my DataMites CDS live project work. The task given was to predict customer churn (whether a customer will leave or not) for a telecom company, understand what drives churn, and create a risk score (CHURN-FLAG) that could be used for retention campaigns. Access to the data was given through a MySQL database instead of a CSV file, so the first step was to connect to the database directly using Python.

## 2. Data Source and a Mismatch I Found

I connected to the MySQL database using SQLAlchemy and pymysql. One small technical issue I ran into: the database password had special characters like ! $ @ & in it, and this broke the connection string at first. I fixed this by URL-encoding the password using Python's `urllib.parse.quote_plus()` before building the connection string.

A bigger issue came up once I actually pulled the data. The project's PDF spec document described a dataset with about 4,617 rows, with columns like State, Account Length, Day/Eve/Night Minutes and Charges, and so on — this looked like older European telecom data. But the real data I got back from the live database was completely different: 243,553 rows and 14 columns, with fields like customer_id, telecom_partner, gender, age, state, city, pincode, date_of_registration, num_dependents, estimated_salary, calls_made, sms_sent, data_used, and churn. This is Indian telecom data (Airtel, BSNL, Reliance Jio, Vodafone), not the European-style data described in the spec.

I checked with `SHOW TABLES;` and confirmed this `telecom_churn_data` table is the only table in the database, so I decided to go ahead with this real data as the correct/authoritative source rather than the outdated PDF description. My guess is that the shared training database was updated at some point after the spec document was originally written, and the documentation was just never updated to match. I'm noting this here because it's a genuine real-world type of problem (documentation not matching the actual data) and it affected how I approached the rest of the project.

## 3. Exploratory Data Analysis (EDA)

Basic checks: the dataset has 243,553 rows and 14 columns, and there are no missing values in any column, which was a nice surprise (no cleaning needed for missing data).

Churn balance: about 80% of customers did not churn (194,726) and about 20% did churn (48,827). This is an imbalanced dataset, so I knew from the start that accuracy alone would not be a fair way to judge any model — I would need to look at precision, recall, and AUC as well.

I then checked each feature against churn, one at a time, to see if any of them showed a real difference between customers who churned and customers who didn't:

| Feature | What I found | Signal? |
|---|---|---|
| telecom_partner | Churn rate ranged only 19.9%–20.4% across Airtel, BSNL, Jio, Vodafone | No |
| gender | 20.3% (F) vs 19.9% (M) — almost identical | No |
| age | Mean age nearly the same for churners vs non-churners (~46 both) | No |
| estimated_salary | Mean salary nearly the same for both groups | No |
| num_dependents | Mean ~2.0 for both groups | No |
| calls_made | Mean ~49 for both groups | No |
| sms_sent | Mean ~24 for both groups | No |
| data_used | Mean ~4,995 for both groups | No |
| tenure_days (derived) | Mean ~1,782 days for both groups | No |
| state | Churn rate ranged 19.4%–21.1% across all 28 states — small, likely noise | No |
| city | Only 6 unique cities; included in modeling for completeness | Checked in model |

This was an important and honestly a bit surprising finding: none of the individual features showed any real difference between customers who churn and customers who don't. Normally in a churn project I would expect at least something like tenure or usage to show some pattern, so I made sure to double- and triple-check this (including the state-level and city-level breakdown) before moving on to modeling, rather than assuming my code was wrong.

## 4. Feature Engineering

Before modeling, I did a few things to prepare the data:

- `date_of_registration` was converted into a new column called `tenure_days` (number of days since registration), since raw dates aren't useful for a model but tenure often is in churn problems.
- `pincode` was dropped. It's stored as a number in the database but it's really a location code, not a true numeric value, so keeping it as a number could mislead a model. Since `state` already covers location and showed no signal, I left `pincode` out to keep things simple.
- `customer_id` and raw `date_of_registration` were dropped, since they're identifiers, not predictive features.
- `telecom_partner`, `gender`, `state`, and `city` were one-hot encoded (turned into 0/1 columns), and the numeric columns were scaled using `StandardScaler`, all inside a single Pipeline, the same approach I used in my capstone projects.

## 5. Modeling

I split the data 80/20 into train and test sets (stratified, so both sets keep the same 80/20 churn balance). I then trained three different models using the same Pipeline structure: Logistic Regression, Random Forest, and XGBoost, all with class-balancing settings (`class_weight='balanced'` or `scale_pos_weight`) since the data is imbalanced.

Results on the test set (AUC is the most reliable metric here, since accuracy can be misleading on imbalanced data):

| Model | AUC | Accuracy |
|---|---|---|
| Logistic Regression | 0.506 | 51% |
| Random Forest | 0.502 | 80% (but predicted 0 churners — see note below) |
| XGBoost | 0.505 | 53% |

An AUC of 0.50 means the model is doing no better than randomly guessing (0.50 = random, 1.00 = perfect). All three models — even though they work in very different ways (linear, bagged trees, boosted trees) — landed at essentially the same AUC of around 0.50. I think this is actually a strong result in its own way: three completely different algorithms agreeing that there's no learnable pattern is more convincing than just one model failing, since it rules out the chance that I picked a bad algorithm.

One thing worth explaining is why Random Forest showed 80% accuracy but 0% recall for churn — this looks good at first glance but it isn't. Since there was no real pattern for it to learn, Random Forest ended up just predicting "No Churn" for almost every customer, which happens to be right 80% of the time simply because 80% of customers really don't churn. Its AUC of 0.502 shows the truth — it wasn't actually able to separate churners from non-churners at all, the high accuracy was just a side effect of the class imbalance. This is a good example of why accuracy alone can be a misleading metric on imbalanced data.

## 6. CHURN-FLAG Risk Score

The project asked for a CHURN-FLAG variable — a risk score that could be used by a retention team to prioritize which customers to reach out to. I used the XGBoost model (best AUC of the three) to generate a churn probability for each customer, and grouped customers into three tiers: High risk (probability ≥ 0.6), Medium risk (0.3–0.6), and Low risk (below 0.3).

**A mistake I made and caught:** My first attempt scored the entire dataset, including the 80% of rows the model had already been trained on. This gave a very different-looking result — High risk customers showed a 54.6% actual churn rate versus just 3.8% for Low risk, which looked like a great result. But this was misleading: the model was partly "remembering" answers for rows it had already seen during training, not genuinely predicting on new data. This is a classic mistake called data leakage.

I fixed this by only scoring the held-out test set — the 20% of data the model never saw during training. This is the correct way to check if a model would actually work in the real world. The corrected result:

| Risk Tier | Customers | Actual churn rate |
|---|---|---|
| High | 1,703 | 19.96% |
| Medium | 45,586 | 20.03% |
| Low | 1,422 | 20.53% |

On the correct, out-of-sample test set, all three risk tiers show almost exactly the same actual churn rate (~20%), which matches the overall churn rate of the dataset. In other words, once measured correctly, the CHURN-FLAG does not actually separate high-risk customers from low-risk customers. This matches the AUC results from Section 5 — everything is consistent once the leakage mistake was fixed.

## 7. Limitations and Why This Happened

The main limitation of this project is that none of the available features — demographics (age, gender, salary, dependents), usage (calls, sms, data), location (state, city), or tenure — show any real relationship with churn in this dataset. I checked this three separate ways (individual feature comparison, three different model types, and risk-tier validation on held-out data) and got the same answer each time, so I'm fairly confident this isn't a mistake on my part, but a real property of this particular dataset.

My best guess is that the churn column in this training database may have been generated separately from the other columns (e.g., randomly assigned for practice purposes), rather than being real customer behavior data. This is common in datasets used for learning/practice, where the label doesn't need to be logically tied to the other columns the way it would be in a real company's data.

In a real business setting, I would expect churn to actually be driven by things that aren't present in this dataset at all, such as: number of complaints or support tickets raised, network quality issues in the customer's area, competitor offers/pricing at the time, billing disputes, or contract type / recent price changes. These would be the kinds of features I'd want to ask for if this were a real client project.

## 8. Conclusion and What I Learned

Even though the churn prediction itself didn't come out strong, I think this project taught me a few real things I wouldn't have learned from a "clean" dataset that just works:

- How to connect Python to a live MySQL database, including handling a password with special characters.
- How to check a project's data against its documentation, rather than assuming the spec is correct.
- How to properly test whether a feature has predictive signal before jumping into modeling.
- Why AUC is a more trustworthy metric than accuracy on imbalanced data, and how a high accuracy number can hide a model that isn't really working (as seen with Random Forest).
- What data leakage looks like in practice, how it can quietly inflate results, and how to catch it by properly separating train and test data when scoring.

Overall, I believe this project is complete and the conclusions are honest and well-supported by the evidence, even though the headline result is "this data doesn't predict churn well" rather than a high-accuracy model. I think that's a more useful and more truthful outcome to report than forcing a good-looking number that wouldn't actually hold up.